### PR TITLE
Remove several functions from collections.js which have ES5 equivalents

### PR DESCRIPTION
**Pull request by Arthri**
## About The Pull Request
This PR should not affect players: it purely serves to clean up some old functions mentioned in https://tgstation13.org/phpBB/viewtopic.php?p=724810.

In summary: Refactor of `collections.ts` and refactor of collection manipulation. The following paragraphs go into the nuances of the PR and reading it is optional.

I originally wanted to remove `flow` from `fp.js`, but I've decided against it until I have a more permanent solution for `sortBy` and `uniq`/`uniqBy`

### `common/collections.ts` Refactor
- `sort(array)` is the same.
- `sortStrings(array)` has been replaced by `sort(array)`.
- Added types to `reduce(...)`.
- Restored old implementation of `map(...)` prior to #60961.
- Rewrote most functions from the form `f(...args)(array)` to `f(array, ...args)`.

### ECMAScript 5 `Array.prototype.*` functions
Most of the functions have direct equivalents in ES5, but behavior varies slightly. Modern JS engines such as V8 and SpiderMonkey(I've not checked for IE) uses optimized objects for array functions such that `typeof [].map(() => {})` is an `object` which is `array`-like, in contrast, the collection functions return an `array`.

All replacement functions are part of ES5([caniuse](https://caniuse.com/es5)), which is supported natively by Internet Explorer version 10 and version 11([MSHTML version](https://en.wikipedia.org/wiki/Trident_(software)) 6 and version 7 respectively).

| `collections.ts` | ES5 |
|-|-|
| `filter((element, index, array) => {})(array)` | `array.filter((element, index, array) => {})` |
| `map(function)(array)` | `array.map(function)` |
| `map((value, key, object) => {})(object)` | * `Object.entries(object).map(([key. value], entries) => {})` |
| `filterMap(array, (element) => {})` | `array.map((element) => {}).filter((element) => element !== undefined)` |
| `reduce((accumulator, currentValue, currentIndex, array) => {}, initialValue)(array)` | `array.reduce((accumulator, currentValue, currentIndex, array) => {}, initialValue)` |
| `zipWith((element1, element2, element3) => {})(array1, array2, array3)` | `zip(array1, array2, array3).map(([element1, element2, element3]) => {})` |
| `sort(array)` | ** `array.sort()` |
| `sortStrings(array)` | ** `array.sort()` |

\*: the second argument of the arrow function is `Object.entries(object)` rather than `object`.
\*\*: there may be some differences in the way these functions sort(such as culture and casing considerations), but they should remain largely the same

### Polyfills

Contrary to the statement on the original forum post, I've not decided to include Lodash because the project also encourages questionable practices. Lodash has a newer competitor, Radash, but I've decided against including that too as it's not mature enough to provide significant benefit over the functions defined in /tg/ui.

`unique` is a function present in core-js(a polyfill for https://github.com/tc39/proposal-array-unique), but I've decided against using it because it lacks TypeScript definitions, and the proposal has been in Stage 1 for quite a while.

### Vectors
Rewrote `vecAdd`, `vecSubtract`, `vecMultiply`, and `vecDivide`. The following compares the current and new implementation of `vecAdd`.

Current Implementation:
Reduce followed by element-wise zip(zip and addition step combined, otherwise the diagram wouldn't fit)

$$
\begin{array}{c c c c}
V_1 & V_2 & V_3 & V_4 \\
\begin{bmatrix}
1\\
2\\
3\end{bmatrix} &
\begin{bmatrix}
4\\
5\\
6\end{bmatrix} &
\begin{bmatrix}
7\\
8\\
9\end{bmatrix} &
\begin{bmatrix}
10\\
11\\
12\end{bmatrix}
\end{array}
\to \begin{array}{c c c}
V_{1 + 2} & V_3 & V_4 \\
\begin{bmatrix}
5\\
7\\
9\end{bmatrix} &
\begin{bmatrix}
7\\
8\\
9\end{bmatrix} &
\begin{bmatrix}
10\\
11\\
12\end{bmatrix}
\end{array}
\to \begin{array}{c c}
V_{1 + 2 + 3} & V_4 \\
\begin{bmatrix}
12\\
15\\
18\end{bmatrix} &
\begin{bmatrix}
10\\
11\\
12\end{bmatrix}
\end{array}
\to \begin{array}{c}
V_R \\
\begin{bmatrix}
22\\
26\\
30\end{bmatrix}
\end{array}
$$

New Implementation:
Zip followed by element-wise reduce

$$
\begin{array}{c c c c}
V_1 & V_2 & V_3 & V_4 \\
\begin{bmatrix}
1\\
2\\
3\end{bmatrix} &
\begin{bmatrix}
4\\
5\\
6\end{bmatrix} &
\begin{bmatrix}
7\\
8\\
9\end{bmatrix} &
\begin{bmatrix}
10\\
11\\
12\end{bmatrix}
\end{array}
\to \begin{array}{c c c}
Z_1 & Z_2 & Z_3\\
\begin{bmatrix}
1\\
4\\
7\\
10\end{bmatrix}
& \begin{bmatrix}
2\\
5\\
8\\
11\end{bmatrix}
& \begin{bmatrix}
3\\
6\\
9\\
12\end{bmatrix}
\end{array}
\to \begin{array}{c}
V_R \\
\begin{bmatrix}
1 + 4 + 7 + 10\\
2 + 5 + 8 + 11\\
3 + 6 + 9 + 12\\
\end{bmatrix}
\end{array}
\to \begin{array}{c}
V_R \\
\begin{bmatrix}
22\\
26\\
30\\
\end{bmatrix}
\end{array}
$$

---

Fixed `vecNormalize`. Previously, it would attempt to divide a vector by a scalar(which is not a valid argument for `vecDivide`). The function has been corrected to properly each vector's component by the scalar.

### Other Changes
- Adjusted some types `NtosNetDownloader.tsx` for brevity.
- Adjusted types around `LibraryAdmin.tsx` to fix type errors.
- Removed `fp.js`' `compose` function.
- Rewrote `StackCrafting.tsx`'s filterRecipeList.

## Why It's Good For The Game
Reduced code debt is good.

### CODE REVIEW

File: tgui/packages/common/vector.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
+/**
```
Comment (jlsnow301) [inline tgui/packages/common/vector.ts:1]: Can you add an explicit return type to these functions please? For compiler efficiency

File: tgui/packages/tgui/components/Chart.tsx [hunk 2]
```diff
@@ -50,11 +50,11 @@ const normalizeData = (
...
 
-  const normalized = map((point: Point) => {
-    return zipWith((value: number, min: number, max: number, scale: number) => {
-      return ((value - min) / (max - min)) * scale;
-    })(point, min, max, scale);
-  })(data);
+  const normalized = data.map((point) =>
+    zip(point, min, max, scale).map(
+      ([value, min, max, scale]) => ((value - min) / (max - min)) * scale,
+    ),
```
Comment (jlsnow301) [inline tgui/packages/tgui/components/Chart.tsx:56]: Is this losing its typing from Point?

**Review by jlsnow301 — COMMENTED** [review]
While I don't disagree with the theme of the pr (we shouldn't have our own versions of the standard lib), I think they were implemented because they handle things in place etc. I would like to hear @stylemistake weigh in on this at least

**Comment by stylemistake** [comment]
You don't need to refactor everything to those functions, only if something lags noticeably on large arrays, so I suggest leave everything as it already is in your PR.

Nested function calls I would say are not that bad, in worst case we can use flow, which I would also redesign to take data first to look nicer (and have stronger typing):

```ts
const output = flow(data, [
  data => map(x => ...),
]);
```

This isn't a priority though, don't put too many things in one PR.

**Comment by stylemistake** [comment]
It can be another PR.

**Comment by stylemistake** [comment]
I noticed that map in collections.js doesn't actually have any performance benefits after @Mothblocks rewrote it, I don't remember why he did it, but maybe something to do with typings.

I know this is a tall order, but I suggest restoring the original code that was there prior to regression, then make them data-first as I described above, then refactor all usages in UIs to use the new format.

I feel this would be the best course of action to preserve the original intent of collections.js (perf optimizations), otherwise there really is no point in functions like filter and map.

Would you be up for doing it, or nah?

**Comment by jlsnow301** [comment]
we're going to be on es2022 soon, will this performance benefit still be realized? enough to make it worth wielding our own standard library?

**Comment by stylemistake** [comment]
> we're going to be on es2022 soon, will this performance benefit still be realized?

Yes.

> enough to make it worth wielding our own standard library?

You decide, but I think 2-3x speedup is significant, especially for UIs that deal with long lists. And I don't advocate for complete replacement of all array methods with these, only where it matters or where they are already used.

**Comment by stylemistake** [comment]
Sure, go ahead!

File: tgui/packages/common/collections.ts [hunk 3]
```diff
@@ -12,33 +12,36 @@
...
-      const result: T[] = [];
-      for (let i = 0; i < collection.length; i++) {
-        const item = collection[i];
-        if (iterateeFn(item, i, collection)) {
-          result.push(item);
-        }
+export const filter = <T>(
+  collection: T[],
+  iterateeFn: (input: T, index: number, collection: T[]) => boolean,
+) => {
```
Comment (jlsnow301) [inline tgui/packages/common/collections.ts:18]: return type here pls

File: tgui/packages/common/collections.ts [hunk 4]
```diff
@@ -293,13 +293,15 @@ const binarySearch = <T, U = unknown>(
...
-  (collection: readonly T[], value: T) => {
-    const copy = [...collection];
-    copy.splice(binarySearch(getKey, collection, value), 0, value);
-    return copy;
-  };
+export const binaryInsertWith = <T, U = unknown>(
+  collection: readonly T[],
+  value: T,
+  getKey: (value: T) => U,
+) => {
```
Comment (jlsnow301) [inline tgui/packages/common/collections.ts:300]: return type

**Pull request merged**